designs/0032-cmdstan-output-formats.md

```diff
@@ -0,0 +1,230 @@
+instead of a single CSV file would make this task easier.
+Another consideration is the use of multi-chain methods; currently multi-chain NUTS-HMC and in the future,
+multi-chain Pathfinder.  When running any kind of ensemble method, we need to record the process/chain_id
+and make this available 
+
+Therefore we need to re-implement the set of output methods used by CmdStan and the core Stan services
+to capture more information in a more structured way.
+At the same time, we need to continue to support the current Stan CSV format in order to support existing downstream processing packages.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+For a given run of an inference algorithm given a model and data,
+the outputs of interest can be categorized in terms of information source and content
+and information structure.
+
+There are three sources of information:
+
+1. Stan program outputs: estimates for all model parameters, transformed parameters, and generated quantities.
+
+2. Inference engine outputs: various diagnostics.
+For the NUTS-HMC sampler, the current outputs are the stepsize and metric, and the per-iteration sampler variables (state).
+The optimization and variational algorithms also report on their respective iterations.
+
+3. Interface-level outputs:
+   + model name, compile options, Stan compiler and (Cmd)Stan version and compile options.
+   + inference algorithm, user-specified configuration options, and default config.
+   + input and output data descriptors.
+   + process-level information:  chain, iteration.
+   + timing information: timestamp information reported by processing events.
+
+There are two general kinds of information structure: tabular and hierarchical.
+[CSV files](https://en.wikipedia.org/wiki/Comma-separated_values) are tabular,
+as are R and Python dataframes and database tables.
+JSON is a [structured data formats](https://en.wikipedia.org/wiki/Data_exchange#Data_exchange_languages)
+which can represent hierarchical structures and arrays.
+
+## Current implementation
+
+We review the way in which the current Stan CSV file is generated for a run of the NUTS-HMC sampler.
+
+### Output mechanism:  callback writers
+
+Outputs are handled by objects of base class `stan::callbacks::writer`
+which is constructed from an output stream.
+Through function call operator overloading, the callback writer class
+methods provide the appropriate formatting for different kinds of data.
+When the user sends a request to the interface to do some kind of inference given a model and data,
+the interfaces first instantiate the Stan model object and the writer callbacks,
+then pass these in to `stan::services` functions which invoke the appropriate algorithm.
+The use of writer callbacks keeps the inference algorithms output-format agnostic.
+The current `stan::services` layer utilities are completely Stan CSV format-centric and will need to be refactored.
```

Comment:
Can you say in what way these are CSV format-centric?  I thought we just output draws as vectors to the callbacks.  That doesn't imply any kind of output at all, much less CSV.  This is evidenced by RStan and PyStan using them with no CSV files in sight.

Reply:
I think that the current set of callbacks and their usage is CSV-centric in at least a couple ways (which I agree should be made explicit in this design doc)

1. The information is all flattened before being passed to these callbacks, so any format is expected to be csv-like in that it only accepts flat rows of data. This is in contrast to something like Arrow/Protobuf/NetCDF/whatever binary format, which generally support things like nested arrays as a native part of their data model. 
2. Anything that is not a draw, e.g. the adaptation information, is passed as pre-formatted string, since the current assumption is that this will just have a `#` prepended to make it a comment in a CSV

Reply:
The `stan::callbacks::stream_writer` adds the requisite CSV formatting, and the CmdStan interface instantiates either `stream_writer` or `unique_stream_writer` objects which are then used everywhere.
updated explanation to make this clear.